flax/experimental/nnx/nnx/nn/lora.py

```
@@ -0,0 +1,192 @@
+    >>> assert wrapper.base_module == linear
+    >>> wrapper.lora_a.value.shape
+    (3, 2)
+    >>> layer.lora_b.value.shape
+    (2, 4)
+    >>> y = layer(jnp.ones((16, 3)))
+    >>> y.shape
+    (16, 4)
+
+  Attributes:
+    in_features: the number of input features.
+    out_features: the number of output features.
+    lora_rank: the rank of the LoRA dimension.
+    base_module: a base module to call and substitute, if possible.
+    dtype: the dtype of the computation (default: infer from input and params).
+    param_dtype: the dtype passed to parameter initializers (default: float32).
+    precision: numerical precision of the computation see `jax.lax.Precision`
+      for details.
+    kernel_init: initializer function for the weight matrices.
+    use_lora_param_type: if yes, LoRA params will be of different Param type.
+      This makes it easier to split model state for trainable/nontrainables.
+  """
+  def __init__(
+    self,
+    in_features: int,
+    out_features: int,
+    lora_rank: int,
+    *,
+    base_module: tp.Optional[nnx.Module] = None,
+    dtype: tp.Optional[Dtype] = None,
+    param_dtype: Dtype = jnp.float32,
+    kernel_init: Initializer = default_kernel_init,
+    lora_param_type: bool = True,
+    rngs: rnglib.Rngs,
+  ):
+    self.in_features = in_features
+    self.out_features = out_features
+    self.dtype = dtype
+    self.param_dtype = param_dtype
+    self.lora_param_type = lora_param_type
+    self.base_module = base_module
+
+    param_type = LoRAParam if self.lora_param_type else nnx.Param
+    self.lora_a = param_type(
+      kernel_init(rngs.params(), (in_features, lora_rank), param_dtype)
+    )
+    self.lora_b = param_type(
+      kernel_init(rngs.params(), (lora_rank, out_features), param_dtype)
+    )
+
+  def __call__(self, x: jax.Array):
+    out = x @ self.lora_a @ self.lora_b
+    if self.base_module is not None:
+      assert callable(self.base_module), "`base_module` must be callable."
```

Lets raise an error here instead

---

Done.